src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2773,24 +2789,51 @@ void ProcessRequiresUnreferencedCode (MethodDefinition method, DependencyKind de
 				return;
 
 			default:
-				// DirectCall, VirtualCall and NewObj are handled by ReflectionMethodBodyScanner
-				// This is necessary since the ReflectionMethodBodyScanner has intrinsic handling for some
-				// of the annotated methods annotated (for example Type.GetType)
-				// and it knows when it's OK and when it needs a warning. In this place we don't know
-				// and would have to warn every time
-
 				// All other cases have the potential of us missing a warning if we don't report it
 				// It is possible that in some cases we may report the same warning twice, but that's better than not reporting it.
 				break;
 			}
 
-			// All override methods should have the same annotations as their base methods (else we will produce warning IL2046.)
-			// When marking override methods with RequiresUnreferencedCode on a type annotated with DynamicallyAccessedMembers,
-			// we should only issue a warning for the base method.
-			if (dependencyKind != DependencyKind.DynamicallyAccessedMember ||
-				!method.IsVirtual ||
-				Annotations.GetBaseMethods (method) == null)
-				CheckAndReportRequiresUnreferencedCode (method);
+			// All override methods should have the same annotations as their base methods
+			// (else we will produce warning IL2046 or IL2092 or some other warning).
+			// When marking override methods via DynamicallyAccessedMembers, we should only issue a warning for the base method.
+			if (dependencyKind == DependencyKind.DynamicallyAccessedMember &&
+				method.IsVirtual &&
+				Annotations.GetBaseMethods (method) != null)
+				return;
+
+			CheckAndReportRequiresUnreferencedCode (method);
+
+			if (_context.Annotations.FlowAnnotations.ShouldWarnWhenAccessedForReflection (method)) {
+				// If the current scope has analysis warnings suppressed, don't generate any
+				if (ShouldSuppressAnalysisWarningsForRequiresUnreferencedCode ())
+					return;
+
+				// ReflectionMethodBodyScanner handles more cases for data flow annotations
+				// so don't warn for those.
+				switch (dependencyKind) {
+				case DependencyKind.AttributeConstructor:
+				case DependencyKind.AttributeProperty:
+					return;
+
+				default:
+					break;
+				}
+
+				// If the method only has annotation on the return value and it's not virtual avoid warning.
+				// Return value annotations are "consumed" by the caller of a method, and as such there is nothing
+				// wrong calling these dynamically. The only problem can happen if something overrides a virtual
+				// method with annotated return value at runtime - in this case the trimmer can't validate
+				// that the method will return only types which fulfill the annotation's requirements.
```

Comment:
I'm not sure I understand why virtual methods are considered specially here. How would you override it at runtime without producing analysis warnings elsewhere, for example when instantiating the overriding type? We can't check annotations on any methods created at runtime so I would expect this to fall into the same bucket.

Reply:
The only way to override a method at runtime is via something like `TypeBuilder` which will need the `MethodInfo` of the base method to override. So we're guarding all the ways to get that `MethodInfo` - which is what this code does. The problem is that the other code in the app may call the base method with its annotation and assume that any type it gets as the return value will fulfill the annotation - but if that value comes from the runtime generated method, linker can't see it and validate - so we warn here.

Non-virtual methods can't be overriden, so there's no way (other than the ref return mentioned above) to change the implementation. Linker checks all implementations to validate that they fulfill the annotation requirements on return values - so no problem there. (Note that interface methods are also considered virtual here and the logic will apply to them as well).

Reply:
It might help to add a blurb like this to the comment:

```csharp
class Foo
{
    [return: DAM(Fields)]
    public abstract Type GetTypeWithFields();
}
```

Then we reflection emit a class that derives from `Foo` and add a virtual method named `GetTypeWithFields` with an implementation that is equivalent to `return SomeOtherType.SomeMethod()`.

Since illink didn't see this, it wouldn't make sure whatever SomeMethod returns meets the requirements of the return type.

Our check ensures reflection.emit won't be able to derive from Foo without raising a warning (Ref.Emit requires the base type be annotated as DAM.All).

Reply:
I added a long description with a sample - hopefully it will make it a bit clearer.

Reply:
We probably didn't do full analysis of this use case, but I would think that getting to a bad place through ref emit will be pretty hard (as in, we've done a good job guarding ourselves). The logic is basically that in order to call any method from a ref emit code, you must be able to reflection access that method (get a MethodInfo). All methods which lead to dangerous places should be annotated and thus should produce warnings when accessed through reflection. So even code which would use reflection from the ref emit code should still cause warnings, since it will have to get MethodInfo of the reflection APIs, which will produce warnings (since those APIs are annotated).

I think we have potential holes around things like out parameters, by ref values and so on - which is basically partially what this discussion is about, but ignoring those cases I can't think of another break (which means absolutely nothing... since I didn't realize the by ref problems either).